sled-agent/src/hardware/illumos/mod.rs

```diff
@@ -73,56 +170,187 @@ fn node_name(subsystem_vid: i32, subsystem_id: i32) -> String {
+                    && l.path()
+                        .file_name()
+                        .map(|f| f.to_string_lossy().ends_with("d0"))
+                        .unwrap_or(false)
+            })
+            .collect::<Vec<_>>();
+
+        if paths.is_empty() {
+            return Err(Error::NoDevLinks(PathBuf::from(devfs_path)));
+        }
+        return Ok(Some(paths[0].path().to_path_buf()));
+    }
+    Ok(None)
+}
+
+// Gather a notion of "device identity".
+fn blkdev_device_identity(devfs_path: &str) -> Result<DiskIdentity, Error> {
+    let path = PathBuf::from(devfs_path);
+    let ctrl_id = nvme::identify_controller(&path)
+        .map_err(|err| Error::MissingNVMEIdentity { path, err })?;
+
+    let vendor =
+        format!("{:x}{:x}", ctrl_id.vendor_id, ctrl_id.subsystem_vendor_id);
+    let serial = ctrl_id.serial;
+    let model = ctrl_id.model;
+    Ok(DiskIdentity { vendor, serial, model })
+}
+
+fn poll_blkdev_node(
+    log: &Logger,
+    disks: &mut HashSet<UnparsedDisk>,
+    mut node: Node<'_>,
+) -> Result<(), Error> {
+    if let Some(driver_name) = node.driver_name() {
+        if driver_name == "blkdev" {
+            let devfs_path = node.devfs_path().map_err(Error::DevInfo)?;
+            let dev_path = get_dev_path_of_whole_disk(&node)?;
+
+            // For some reason, libdevfs doesn't prepend "/devices" when
+            // referring to the path, but it still returns an absolute path.
+            //
+            // Validate that we're still using this leading slash, but make the
+            // path "real".
+            assert!(devfs_path.starts_with('/'));
+            let devfs_path = format!("/devices{devfs_path}");
+            let mut device_id = None;
+            while let Some(parent) = node.parent().map_err(Error::DevInfo)? {
+                node = parent;
+
+                if let Some(driver_name) = node.driver_name() {
+                    if driver_name == "nvme" {
+                        device_id = Some(blkdev_device_identity(&format!(
+                            "/devices{}:devctl",
+                            node.devfs_path().map_err(Error::DevInfo)?
+                        ))?);
+                    }
+                }
+                let slot_prop_name = "physical-slot#".to_string();
+                if let Some(Ok(slot_prop)) =
+                    node.props().into_iter().find(|prop| {
+                        if let Ok(prop) = prop {
+                            prop.name() == slot_prop_name
+                        } else {
+                            false
+                        }
+                    })
+                {
+                    let slot = slot_prop.as_i64().ok_or_else(|| {
+                        Error::UnexpectedPropertyType {
+                            name: slot_prop_name,
+                            ty: "i64".to_string(),
+                        }
+                    })?;
+
+                    let variant = if let Some(variant) =
```

Comment:
I think this could also use `let .. else`:

```rust
let Some(variant) = slot_to_disk_variant(slot) else {
    warn!(log, "Slot# {slot} is not recognized as a disk: {devfs_path}");
    break;
};
```

Reply:
Updated